src/voter/voting_round.rs

```diff
@@ -232,12 +233,45 @@ impl<H, N, E: Environment<H, N>> VotingRound<H, N, E> where
 						self.env.precommit_equivocation(self.votes.number(), e);
 					}
 				}
+				Message::Primary(primary) => {
+					let primary_index = self.votes.number() as usize % self.votes.voters().voters.len();
+					let primary_id = self.votes.voters().voters.get(primary_index)
+						.expect("primary_index in range by definition; qed").0.clone();
+					if id == primary_id {
+						self.primary_block = Some((primary.target_hash, primary.target_number));
+					}
+				}
 			};
 		}
 
 		Ok(())
 	}
 
+	fn primary(&mut self, last_round_state: &RoundState<H, N>) -> Result<(), E::Error> {
+		match self.state {
+			Some(State::Start(_, _)) => {
+				let last_round_estimate = last_round_state.estimate.clone()
+					.expect("Rounds only started when prior round completable; qed");
+				let last_round_finalized = last_round_state.finalized.clone()
+					.expect("Rounds only started when prior round completable; qed"); // TODO: check this.
```

Comment:
these `expect`s are pretty subtle. Maybe it's better to log and just not propose if this happens?